include/aws/io/logging.h

```diff
@@ -0,0 +1,175 @@
+ */
+enum aws_log_level {
+    AWS_LL_NONE = AWS_LOG_LEVEL_NONE,
+    AWS_LL_FATAL = AWS_LOG_LEVEL_FATAL,
+    AWS_LL_ERROR = AWS_LOG_LEVEL_ERROR,
+    AWS_LL_WARN = AWS_LOG_LEVEL_WARN,
+    AWS_LL_INFO = AWS_LOG_LEVEL_INFO,
+    AWS_LL_DEBUG = AWS_LOG_LEVEL_DEBUG,
+    AWS_LL_TRACE = AWS_LOG_LEVEL_TRACE,
+
+    AWS_LL_COUNT
+};
+
+/**
+ * Log subject is a to-be-defined key, indendent of level, that will allow for additional filtering options.
+ *
+ * The general idea is to support the logging of related functionality and allow for level control with a
+ * finer-grained approach.
+ *
+ * For example, enable TRACE logging for mqtt-related log statements, but only WARN logging everywhere else.
+ */
+typedef uint64_t aws_log_subject_t;
+
+#define AWS_LOG_SUBJECT_NONE ((aws_log_subject_t)0)
+
+struct aws_logger;
+
+/**
+ * We separate the log level function from the log call itself so that we can do the filter check in the macros (see below)
+ *
+ * By doing so, we make it so that the variadic format arguments are not even evaluated if the filter check does not succeed.
+ */
+struct aws_logger_vtable {
+    int(*log_fn)(struct aws_logger *logger, enum aws_log_level log_level, aws_log_subject_t subject, const char *format, ...);
+    enum aws_log_level(*get_log_level_fn)(struct aws_logger *logger, aws_log_subject_t subject);
+    int(*cleanup_fn)(struct aws_logger *logger);
+};
+
+struct aws_logger {
+    struct aws_logger_vtable *vtable;
+    struct aws_allocator *allocator;
+    void *p_impl;
+};
+
+AWS_EXTERN_C_BEGIN
+
+/**
+ * Sets the aws logger used globally across the process.  Not thread-safe.  Must only be called once.
+ */
+AWS_COMMON_API
+void aws_logging_set(struct aws_logger *logger);
+
+/**
+ * Gets the aws logger used globally across the process.
+ */
+AWS_COMMON_API
+struct aws_logger *aws_logging_get(void);
+
+/**
+ * Cleans up all resources used by the logger; simply invokes the cleanup v-function
+ */
+AWS_COMMON_API
+int aws_logger_cleanup(struct aws_logger *logger);
+
+/**
+ * Converts a log level to a c-string constant.  Intended primarily to support building log lines that
+ * include the level in them, i.e.
+ *
+ * [ERROR] 10:34:54.642 01-31-19 - Json parse error....
+ */
+AWS_COMMON_API
+int aws_logging_log_level_to_string(enum aws_log_level log_level, const char **level_string);
+
+AWS_EXTERN_C_END
+
+/**
+ * The base formatted logging macro that all other formatted logging macros resolve to.
+ * Checks for a logger and filters based on log level.
+ *
+ */
+#define AWS_LOGF_RAW(log_level, subject, ...)                                                                   \
+{                                                                                                               \
+    struct aws_logger *logger = aws_logging_get();                                                              \
+    if (logger != NULL && log_level > 0 && (logger->vtable->get_log_level_fn)(logger, subject) >= log_level) {  \
```

Comment:
If you re-order this so that the logger fetch/assignment is in the if, you can check log_level > 0 first and early out before fetching the logger.

Reply:
Good point, in fact, I should turn the log_level >0 check into an initial assert.